rule/fetcher_default_test.go

```diff
@@ -32,62 +33,89 @@ import (
 )
 
 const testRule = `[{"id":"test-rule-5","upstream":{"preserve_host":true,"strip_path":"/api","url":"mybackend.com/api"},"match":{"url":"myproxy.com/api","methods":["GET","POST"]},"authenticators":[{"handler":"noop"},{"handler":"anonymous"}],"authorizer":{"handler":"allow"},"mutators":[{"handler":"noop"}]}]`
+const testConfigPath = "../test/update"
+
+func copyToFile(t *testing.T, src string, dst *os.File) {
+	t.Helper()
+
+	source, err := os.Open(filepath.Join(testConfigPath, src))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer source.Close()
+
+	_, err = dst.Seek(0, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = dst.Truncate(0)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = io.Copy(dst, source)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
 
 func TestFetcherReload(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	viper.Reset()
-	conf := internal.NewConfigurationWithDefaults() // this resets viper and must be at the top
+
+	configFile, err := os.CreateTemp(t.TempDir(), "config-*.yaml")
+	require.NoError(t, err)
+	t.Cleanup(func() { configFile.Close() })
+
+	configChanged := make(chan struct{})
+
+	conf := internal.NewConfigurationWithDefaults(
+		configx.WithContext(ctx),
+		configx.WithLogger(logrusx.New("", "", logrusx.ForceLevel(logrus.TraceLevel))),
+		configx.SkipValidation(),
+		configx.WithConfigFiles(configFile.Name()),
+		configx.AttachWatcher(func(event watcherx.Event, err error) {
+			go func() { configChanged <- struct{}{} }()
+		}),
+	)
 	r := internal.NewRegistry(conf)
-	testConfigPath := "../test/update"
 
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
 		_, _ = w.Write([]byte(testRule))
 	}))
 	defer ts.Close()
 
-	tempdir := os.TempDir()
-
-	id := uuid.New().String()
-	configFile := filepath.Join(tempdir, ".oathkeeper-"+id+".yml")
-	require.NoError(t, ioutil.WriteFile(configFile, []byte(""), 0666))
-
-	l := logrusx.New("", "", logrusx.ForceLevel(logrus.TraceLevel))
-	viperx.InitializeConfig("oathkeeper-"+id, tempdir, nil)
-	viperx.WatchConfig(l, nil)
-
-	go func() {
-		require.NoError(t, r.RuleFetcher().Watch(ctx))
-	}()
+	go func() { require.NoError(t, r.RuleFetcher().Watch(ctx)) }()
 
 	// initial config without a repo and without a matching strategy
-	config, err := ioutil.ReadFile(path.Join(testConfigPath, "config_no_repo.yaml"))
-	require.NoError(t, err)
-	require.NoError(t, ioutil.WriteFile(configFile, config, 0666))
+	copyToFile(t, "config_no_repo.yaml", configFile)
+	<-configChanged
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Do we need these sleeps? And if so, why is it 100ms? Would this be flaky on a slower system?

Reply:
Hmm, now with the syncs these are probably not needed any more. Runs without flakes locally, will see in the CI.